8322630: Remove ICStubs and related safepoints

fisk opened the pull request:
ICStubs solve an atomicity problem when setting both the destination and data of an inline cache. Unfortunately, it also leads to occasional safepoint carpets when multiple threads need to ICRefill the stubs at the same time, and spurious GuaranteedSafepointInterval "Cleanup" safepoints every second. This patch changes inline caches to not change the data part at all during the nmethod life cycle, hence removing the need for ICStubs.

The new scheme is less stateful. Instead of adding and removing callsite metadata back and forth when transitioning inline cache states, it installs all state any shape of call will ever need at resolution time in a struct that I call CompiledICData. This reduces inline cache state changes to simply changing the destination of the call, and it doesn't really matter what state transitions to what other state.

With this patch, we get rid of ICStub and ICBuffer classes and the related ICRefill and almost all Cleanup safepoints in practice. It also makes the inline cache code much simpler.

I have tested the changes from tier1-7, and run through full aurora performance tests.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322630](https://bugs.openjdk.org/browse/JDK-8322630): Remove ICStubs and related safepoints (**Enhancement** - P4)


### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**) ⚠️ Review applies to [6dd64b50](https://git.openjdk.org/jdk/pull/17495/files/6dd64b5075b1b4c5dcc28fd78fa45002d1306b3a)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [29790afe](https://git.openjdk.org/jdk/pull/17495/files/29790afee9f24e00fb7b8dec53e03e0b5c283059)


### Contributors
 * Martin Doerr `<mdoerr@openjdk.org>`
 * Aleksey Shipilev `<shade@openjdk.org>`
 * Amit Kumar `<amitkumar@openjdk.org>`
 * Robbin Ehn `<rehn@openjdk.org>`
 * Aleksei Voitylov `<avoitylov@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17495/head:pull/17495` \
`$ git checkout pull/17495`

Update a local copy of the PR: \
`$ git checkout pull/17495` \
`$ git pull https://git.openjdk.org/jdk.git pull/17495/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17495`

View PR using the GUI difftool: \
`$ git pr show -t 17495`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17495.diff">https://git.openjdk.org/jdk/pull/17495.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17495#issuecomment-1900580988)

fisk commented:
/contributor add @TheRealMDoerr

fisk commented:
/contributor add @shipilev

fisk commented:
/contributor add @offamitkumar

fisk commented:
/contributor add @robehn

fisk commented:
/contributor add avoitylov

fisk commented:
Thanks to the OpenJDK port maintainers for picking this up! All added to contributors, hopefully.

dougxc commented on src/hotspot/share/runtime/vmStructs.cpp:
```diff
@@ -211,8 +211,10 @@
   volatile_nonstatic_field(ArrayKlass,         _lower_dimension,                              ArrayKlass*)                           \
-  nonstatic_field(CompiledICHolder,            _holder_metadata,                              Metadata*)                             \
-  nonstatic_field(CompiledICHolder,            _holder_klass,                                 Klass*)                                \
+  volatile_nonstatic_field(CompiledICData,     _speculated_method,                            Method*)                               \
```
Please duplicate these `CompiledICData` declarations in `vmStructs_jvmci.cpp` so that they can be used by Graal to ascertain whether ICStubs are in use (Graal is still supporting multi JDK versions).

voitylov commented:
ARM32 passed relevant tests after the update. Thanks @fisk !

TheRealMDoerr commented:
Test results are good on all SAP supported platforms (CPU: x86_64, aarch64, PPC64; OS: linux, Windows, AIX). Performance looks also good.

robehn commented:
I believe this was the last major piece in task we started over 5y ago of removing runtime safepoint and latencies.
Or as some might say, we finally have a runtime good enough to run ZGC ;) (and Shenandoah).
Thank you @fisk for completing this milestone!

Risc-v passes my testing. (vf2 (t1) + qemu (t1-2), ran twice, once on v3 branch and once this pr)

tschatzl commented:
Fwiw, the change makes class unloading regress significantly in a class unloading stress test (unloading 60k classes), seemingly tripling the time it takes for the "Purge Unlinked NMethods" phase (~20ms -> ~60ms).

This may not be a problem for the concurrent gcs, but can be for the STW ones.

fisk commented:
> Fwiw, the change makes class unloading regress significantly in a class unloading stress test (unloading 60k classes), seemingly tripling the time it takes for the "Purge Unlinked NMethods" phase (~20ms -> ~60ms).
> 
> This may not be a problem for the concurrent gcs, but can be for the STW ones.
> 
> (Overall max G1 remark pause times went from 160ms to 220ms, regular Remark pauses which do class unloading from 120ms to 160ms).

If the level of ~1 ms per 1000 unloaded classes worth of latency issue is crucial for G1, then maybe it's time for G1 to at least run purging concurrently?

TheRealMDoerr commented:
Do we know what makes that slower? Is it the glibc overhead of `delete ic->data();`?

fisk commented:
@tschatzl what program did you run, so I can reproduce?

tschatzl commented:
I think the `ccstress` application attached to https://bugs.openjdk.org/browse/JDK-8315503 should show the issue (not tested - today I've been busy completing a JDK 22 bugfix).

The actual application I'm using is specjbb2015 that basically executes that ccstress application as a java agent in parallel to jbb2015 just for extra load. I can certainly give you that.

Tomorrow I will also spend some time investigating this change a bit more (also on aarch64 - I think other factors may easily outweigh this difference), but I think that @TheRealMDoerr is correct about the C heap allocator just being very slow.
Glibc being slow is a "known" issue, other `delete` calls in class unloading already take a significant chunk of that phase, so there is already some interest (from me) to do something about them.

There are ideas how to handle this, one of them being moving this (and other) `delete` calls into some concurrent phase somehow (in addition to making metaspace purging concurrent). That would obviously only help G1 though, so maybe there is some better option.

TheRealMDoerr commented:
On linux, the time for "Purge Unlinked NMethods" goes down when I comment out `delete ic->data();` and ignore the memory leak. (MacOS seems to be ok with it.)
Adding trace code to `purge_ic_callsites` shows that we often have 0 or 2 ICData instances, sometimes up to 30 ones.
It would be good to think a bit about the allocation scheme. Some ideas would be

- Allocate ICData in an array per nmethod instead of individually. That should help to some degree and also improve data locality (and hence cache efficiency). Would also save iterating over the relocations. It's not very complex.
- Instead of freeing ICData instances, we could enqueue them and either reuse or free them during a concurrent phase. This may be a bit complicated. Not sure if it's worth it.

tschatzl commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -1463,10 +1466,11 @@ void nmethod::purge(bool free_code_cache_data, bool unregister_nmethod) {
     ec = next;
   }
 
+  purge_ic_callsites();
+
```
(Github does not allow me to attach this comment to the correct place):
At the start of this method, there is some comment about
```
    // Already unlinked. It can be invoked twice because concurrent code cache
    // unloading might need to restart when inline cache cleaning fails due to
    // running out of ICStubs, which can only be refilled at safepoints
```
This comment and the whole mechanism to prevent this may be outdated since there are no ICStubs and the associated safepoints any more; maybe it is worth keeping the flag to provide an assert though?
I did not check the code flow yet, just going from the comment.

tschatzl replied:
I think the flag is still required, just the comment needs to be fixed then.

fisk replied:
Good point! I actually think we might not need it any more. I added the _is_unlinked field because of ICStubs causing inline cache cleaning from a concurrent GC to fail because going to the cleaned state requires an ICStub and we are out of memory for it, requiring the GC to request a safepoint and then restart the code cache unloading. If there is still a separate reason why we might call unlink twice on the same nmethod, I don't know why that is. Maybe something new in G1? Having said that, I don't mind keeping the guard to make it more robust. What do you think? Either way, I'll remove the inaccurate comment.

tschatzl commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -1219,29 +1223,29 @@ void nmethod::verify_clean_inline_caches() {
+        CodeBlob *cb = CodeCache::find_blob(ic->destination());
         assert(cb != nullptr, "destination not in CodeBlob?");
         nmethod* nm = cb->as_nmethod_or_null();
         if( nm != nullptr ) {
```
Maybe fix this while in the area similar to other places.

fisk replied:
Sorry, I'm not sure I understand what you would like me to change here.

tschatzl replied:
```suggestion
        if (nm != nullptr ) {
```
Sorry.

tschatzl reviewed (changes requested):
Something I found while looking a bit at the code.

xmas92 commented on src/hotspot/cpu/arm/compiledIC_arm.cpp:
```diff
@@ -87,22 +86,22 @@ address CompiledStaticCall::emit_to_interp_stub(CodeBuffer &cbuf, address mark)
+void CompiledDirectCall::set_to_interpreted(const methodHandle& callee, address entry) {
   address stub = find_stub();
   guarantee(stub != nullptr, "stub not found");
 
```
The other platforms removed the trace logging here. If the ARM porters still want this in at least update to log the correct class name. `s/CompiledDirectStaticCall/CompiledDirectCall/`

voitylov replied:
fixed by removing the trace logging.

xmas92 commented on src/hotspot/cpu/arm/sharedRuntime_arm.cpp:
```diff
@@ -628,10 +627,9 @@ AdapterHandlerEntry* SharedRuntime::generate_i2c2i_adapters(MacroAssembler *masm
 
-  __ load_klass(receiver_klass, receiver);
-  __ ldr(holder_klass, Address(Ricklass, CompiledICHolder::holder_klass_offset()));
-  __ ldr(Rmethod, Address(Ricklass, CompiledICHolder::holder_metadata_offset()));
-  __ cmp(receiver_klass, holder_klass);
+  __ ic_check(1 /* end_alignment */);
+  __ ldr(Rmethod, Address(receiver_klass, CompiledICData::speculated_method_offset()));
```
Maybe I am missing something here but this looks very wrong. The speculated `Klass*` gets loaded into `R4` (which `receiver_klass` alias) in `ic_check` this load would result in loading a `InstanceKlass*` c++ vtable pointer. 
`Ricklass` (`R8` alias) contains the  `CompiledICData*` .

I would think the correct diff would be
```diff
-  const Register receiver_klass = R4;
-
-  __ load_klass(receiver_klass, receiver);
-  __ ldr(holder_klass, Address(Ricklass, CompiledICHolder::holder_klass_offset()));
-  __ ldr(Rmethod, Address(Ricklass, CompiledICHolder::holder_metadata_offset()));
-  __ cmp(receiver_klass, holder_klass);
+  __ ic_check(1 /* end_alignment */);
+  __ ldr(Rmethod, Address(Ricklass, CompiledICData::speculated_method_offset()));
```

The fact that you say ARM32 tests are passing makes me doubt my understanding of the inline cache.

voitylov replied:
hotspot jtreg tests pass in both variants, but you finding is correct. Updated the code with your suggestion.

xmas92 commented on src/hotspot/share/code/compiledIC.cpp:
```diff
@@ -75,593 +67,337 @@ bool CompiledICLocker::is_safe(address code) {
+  c_ic->verify();
+  return c_ic;
 }
 
-// Clears the IC stub if the compiled IC is in transition state
-void CompiledIC::clear_ic_stub() {
-  if (is_in_transition_state()) {
-    ICStub* stub = ICStub::from_destination_address(stub_address());
-    stub->clear();
-  }
+CompiledIC* CompiledIC_at(CompiledMethod* nm, address call_site) {
+  RelocIterator iter(nm, call_site, call_site + 1);
+  iter.next();
+  CompiledIC* c_ic = new CompiledIC(&iter);
+  c_ic->verify();
+  return c_ic;
 }
 
-//-----------------------------------------------------------------------------
-// High-level access to an inline cache. Guaranteed to be MT-safe.
+CompiledIC* CompiledIC_at(Relocation* call_reloc) {
+  address call_site = call_reloc->addr();
+  CompiledMethod* cm = CodeCache::find_blob(call_reloc->addr())->as_compiled_method();
+  RelocIterator iter(cm, call_site, call_site + 1);
+  iter.next();
+  CompiledIC* c_ic = new CompiledIC(&iter);
+  c_ic->verify();
+  return c_ic;
+}
```
Purely a style thing but could rewrite the `CompiledIC* CompiledIC_X(...)` functions in terms of each other. Made them all fit very well even on my small laptop screen.
```c++

CompiledIC* CompiledIC_before(CompiledMethod* nm, address return_addr) {
  address call_site = nativeCall_before(return_addr)->instruction_address();
  return CompiledIC_at(nm, call_site);
}

CompiledIC* CompiledIC_at(CompiledMethod* nm, address call_site) {
  RelocIterator iter(nm, call_site, call_site + 1);
  iter.next();
  return CompiledIC_at(&iter);
}

CompiledIC* CompiledIC_at(Relocation* call_reloc) {
  address call_site = call_reloc->addr();
  CompiledMethod* cm = CodeCache::find_blob(call_reloc->addr())->as_compiled_method();
  return CompiledIC_at(cm, call_site);
}

CompiledIC* CompiledIC_at(RelocIterator* reloc_iter) {
  CompiledIC* c_ic = new CompiledIC(reloc_iter);
  c_ic->verify();
  return c_ic;
}
```

xmas92 commented on src/hotspot/share/code/compiledMethod.hpp:
```diff
@@ -373,13 +373,12 @@ class CompiledMethod : public CodeBlob {
   // Execute nmethod barrier code, as if entering through nmethod call.
   void run_nmethod_entry_barrier();
 
   // Verify and count cached icholder relocations.
```
The comment belonged to the removed method.

xmas92 commented on src/hotspot/share/code/compiledIC.cpp:
```diff
@@ -75,593 +67,337 @@ bool CompiledICLocker::is_safe(address code) {
-  if (cb->is_adapter_blob()) {
-    return true;
-  } else if (cb->is_vtable_blob()) {
-    return VtableStubs::is_icholder_entry(entry);
```
`VtableStubs::is_icholder_entry` is no longer used. Should be removed as well.

xmas92 commented on src/hotspot/share/oops/compiledICHolder.hpp:
```diff
@@ -1,92 +0,0 @@
-//
-
-
-class CompiledICHolder : public CHeapObj<mtCompiler> {
```
Still has a forward declaration in `src/hotspot/share/oops/oopsHierarchy.hpp`

xmas92 commented on src/hotspot/share/runtime/vmStructs.cpp:
```diff
@@ -211,8 +211,10 @@
   nonstatic_field(ArrayKlass,                  _dimension,                                    int)                                   \
   volatile_nonstatic_field(ArrayKlass,         _higher_dimension,                             ObjArrayKlass*)                        \
   volatile_nonstatic_field(ArrayKlass,         _lower_dimension,                              ArrayKlass*)                           \
-  nonstatic_field(CompiledICHolder,            _holder_metadata,                              Metadata*)                             \
-  nonstatic_field(CompiledICHolder,            _holder_klass,                                 Klass*)                                \
```
Making the serviceability agent aware of CompiledICData seems like a RFE. However CompiledICHolder has a mirror in the SA. Should probably nuke `src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/CompiledICHolder.java`

xmas92 reviewed (commented):
Just did an initial read through of the PR. Just added some cleanup suggestion. Also noticed something I though looked wrong in the ARM32 port. 

I also went through and tried to find the handful of places in the codebase where the term `ICHolder` (or its derivatives) were still used.  Put them in a separate branch to not clutter this PR. Would be nice to take this all the way and not have stale comments or naming lurking about.  (Also nuked the `DECC` copy-paste-typo)
Comment cleanups:
f1bb02ea472eb314c93d80b830c59bd03e280116

All platforms use `data` as a register alias for the `CompileICData*` register in the `ic_check`. But c2i and itable stubs still use `holder`.  Maybe go all the way here?
5422ed32def491bd1e145959b7f3c49c88cfc50e

Also for PPC and s390 I think the code is easier to understand if the global inline cache register aliases these platforms have are used. But maybe that is just me.
39c0a7ede5187cba52d6fcf48c0852213c48c899

As for the implementation I could not see anything wrong (except the ARM32 port). But I'll leave it people with more expertise in this area.

voitylov commented:
I'll check the ARM32 part of @xmas92 comments early next week.

robehn commented:
> > I believe this was the last major piece in task we started over 5y ago of removing runtime safepoint and latencies. Or as some might say, we finally have a runtime good enough to run ZGC ;) (and Shenandoah). Thank you @fisk for completing this milestone!
> > Risc-v passes my testing. (vf2 (t1) + qemu (t1-2), ran twice, once on v3 branch and once this pr)
> > @RealFYang can you please review RV changes?
> 
> @robehn @fisk : I went through the RV changes and I saw small code cleanup could be done. [rv-extra-cleanup.diff.txt](https://github.com/openjdk/jdk/files/14079614/rv-extra-cleanup.diff.txt)
> 
> BTW: I also performed tier1-3 and hotspot:tier4 tests on hifive unmatched board, result is good.

Thank you @RealFYang !

@fisk please apply [rv-extra-cleanup.diff.txt](https://github.com/openjdk/jdk/files/14079614/rv-extra-cleanup.diff.txt)

tschatzl commented:
> On linux, the time for "Purge Unlinked NMethods" goes down when I comment out delete ic->data(); and ignore the memory leak. (MacOS seems to be ok with it.)
>Adding trace code to purge_ic_callsites shows that we often have 0 or 2 ICData instances, sometimes up to 30 ones.
>It would be good to think a bit about the allocation scheme. Some ideas would be

>    Allocate ICData in an array per nmethod instead of individually. That should help to some degree and also improve data locality (and hence cache efficiency). Would also save iterating over the relocations when purging unlinked NMethods. It's not very complex.
>    Instead of freeing ICData instances, we could enqueue them and either reuse or free them during a concurrent phase. This may be a bit complicated. Not sure if it's worth it.
>    Allocate in Metaspace?

Sorry for being unresponsive for a bit.

Yes, the issue is the new `delete ic->data()`; but also the iteration over the relocinfo here is almost as expensive in my tests.

So the idea to allocate ICData in a per nmethod basis (and actually some other existing C heap allocations that are also `delete`d in this phase) seems the most promising to me.

Also came up with the other suggestions, but I think that first one seems best to me at first glance. I did not really like the second because enqueuing adds another indirection for first gathering all of them and then separately free them.

Metaspace is something I do not know that well to comment on that option.

I am open to moving this improvement, if it is not easy to do, into a separate CR.

fisk commented:
Thanks for the reviews! I applied the cleanups from @xmas92 and @RealFYang. I added the JVMCI hook for @dougxc and started bulk allocating and freeing CompiledICData to deal with the situation reported by @tschatzl. I haven't touched the ARM32 code though - waiting for @voitylov there.

dougxc commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -1100,7 +1134,12 @@ void MacroAssembler::get_vm_result_2(Register metadata_result, Register java_thr
+  align(modulus, offset());
+}
+
+void MacroAssembler::align(int modulus, int target) {
```
It would be nice to document what this extra `align` function does.

fisk replied:
Good idea. I wrote a comment.

fisk commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -995,10 +996,43 @@ address MacroAssembler::ic_call(address entry, jint method_index) {
 }
 
+int MacroAssembler::ic_check_size() {
+  return NativeInstruction::instruction_size * 7;
```
Good point. I sort of assumed we will always get a far jump, but maybe I'm wrong. Either way, I added a check if the target is far or not, so we can select 5 or 7 accordingly.

voitylov commented:
@fisk Sorry for the delay. I prepared a patch that addresses the review comments for ARM32.

[jdk_ic_review_comments.patch](https://github.com/openjdk/jdk/files/14094946/jdk_ic_review_comments.patch)

fisk commented:
> @fisk Sorry for the delay. I prepared a patch that addresses the review comments for ARM32.
> 
> [jdk_ic_review_comments.patch](https://github.com/openjdk/jdk/files/14094946/jdk_ic_review_comments.patch)

Thanks for the fix. I uploaded it to the PR.

dougxc commented:
The PR to adapt Graal for this change: https://github.com/oracle/graal/pull/8287

TheRealMDoerr commented:
Thanks for the improvements! Tests are still passing on SAP supported platforms.

fisk commented:
> Thanks for the improvements! Tests are still passing on SAP supported platforms.

Thank you for running through your tests!

fisk commented:
> All my comments have been addressed. 
> 
> As previously mentioned as long as the performance is there, then this looks good.

Thanks for the review, @xmas92. Any other takers?

dean-long commented:
I saw an earlier version, but I plan to look at your latest, soon.

fisk commented:
> I saw an earlier version, but I plan to look at your latest, soon.

Thank you @dean-long I appreciate it.

dean-long commented on src/hotspot/share/opto/output.cpp:
```diff
@@ -3412,6 +3412,9 @@ void PhaseOutput::install_code(ciMethod*         target,
       _code_offsets.set_value(CodeOffsets::OSR_Entry, _first_block_size);
     } else {
+      if (!target->is_static()) {
+        _code_offsets.set_value(CodeOffsets::Entry, _first_block_size - MacroAssembler::ic_check_size());
```
This looks tricky. I think it means CodeOffsets::Entry starts after the alignment padding NOPs. If that's true then the `ic_check` functions could use a comment explaining that alignment needs to come first, not last. A comment here wouldn't hurt either.

fisk replied:
Yes that's exactly it. I found that I got a less fortunate nop encoding that actually did show up as a tiny regression. It was fixed by not running the nops. I'll write a comment.

dean-long commented on src/hotspot/cpu/aarch64/aarch64.ad:
```diff
@@ -2221,14 +2221,7 @@ void MachUEPNode::emit(CodeBuffer& cbuf, PhaseRegAlloc* ra_) const
-  __ br(Assembler::EQ, skip);
-  __ far_jump(RuntimeAddress(SharedRuntime::get_ic_miss_stub()));
-  __ bind(skip);
+  __ ic_check(CodeEntryAlignment);
```
I'm not sure we want to increase the alignement to CodeEntryAlignment here.  I believe C2 already aligns the root block to CodeEntryAlignment.  @theRealAph, what do you think?

fisk replied:
I can change it to InteriorEntryAlignment which is what the VEP block is aligned to. I think it's still good that the root block is aligned to CodeEntryAlignment though. I heard some HW sheds some tears when their instruction cache lines also contain data that doesn't correspond to any actual instructions. But we can definitely align to InteriorEntryAlignment here.

robehn replied:
To add to what Erik is saying here, we have notice that some CPU's have L1I and L1D exclusive, a cache-line can just be in of them. So you can end up with it bouncing between them.

Secondly, if an I-fetch gets data not intended to be decoded, the cpu can get upset in some cases.

dean-long replied:
> Secondly, if an I-fetch gets data not intended to be decoded, the cpu can get upset in some cases.

I've heard that in the past, and always wondered, does it matter if the data is "reachable"?  What if it's on the other side of an unconditional branch?

robehn replied:
> > Secondly, if an I-fetch gets data not intended to be decoded, the cpu can get upset in some cases.
> 
> I've heard that in the past, and always wondered, does it matter if the data is "reachable"? What if it's on the other side of an unconditional branch?

I don't think it's an guarantee. E.g. an uncond branch may be encoded with multiple instruction, so it might be seen as an indirect jump.

dean-long commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -995,10 +996,47 @@ address MacroAssembler::ic_call(address entry, jint method_index) {
+  Register data = rscratch2;
+  Register tmp1 = rscratch1;
+  Register tmp2 = r10;
+
```
It would be nice if we could still call `verify_oop(receiver)` here, but I see that would complicate `ic_check_size()`.

fisk replied:
Yeah I thought the same. As it's important for correctness that ic_check_size is accurate, I was hoping to have as few different modes in it as possible.

tschatzl commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -2284,15 +2291,23 @@ void nmethod::verify() {
   if (!is_not_installed()) {
     if (CompiledICLocker::is_safe(this)) {
-      CompiledIC_at(this, call_site);
+      if (is_inline_cache) {
+        CompiledIC_at(this, call_site);
+      } else {
+        CompiledDirectCall::at(call_site);
+      }
     } else {
       CompiledICLocker ml_verify(this);
-      CompiledIC_at(this, call_site);
+      if (is_inline_cache) {
+        CompiledIC_at(this, call_site);
+      } else {
+        CompiledDirectCall::at(call_site);
+      }
```
Since `CompiledICLocker` does not lock anyway if it is safe (using the `is_safe` method), isn't the `if` superfluous here (and just the `else` part do the same)?

fisk replied:
That is true for DefaultICProtectionBehaviour::lock, but not ZGC and Shenandoah.

tschatzl commented on src/hotspot/share/code/compiledMethod.cpp:
```diff
@@ -466,70 +428,33 @@ class CheckClass : public MetadataClosure {
+static void clean_if_nmethod_is_unloaded(CompiledDirectCall *cdc, CompiledMethod* from,
                                          bool clean_all) {
-  return clean_if_nmethod_is_unloaded(csc, csc->destination(), from, clean_all);
+  CodeBlob* cb = CodeCache::find_blob(cdc->destination());
+  if (!cb->is_compiled()) {
+    return;
+  }
+  CompiledMethod* cm = cb->as_compiled_method();
+  if (clean_all || !cm->is_in_use() || cm->is_unloading() || cm->method()->code() != cm) {
+    cdc->set_to_clean();
+  }
```
Maybe a single `clean_if_nmethod_is_unloaded()` with the destination address as a parameter would avoid the code duplication with the other variant; otherwise another static helper function would be great.

fisk replied:
Good point. I'll try on a template thing.

tschatzl commented on src/hotspot/share/code/compiledIC.hpp:
```diff
@@ -77,237 +54,105 @@ class CompiledICLocker: public StackObj {
-  bool    _is_optimized;       // it is an optimized virtual call (i.e., can be statically bound)
-  bool    _to_interpreter;     // Call it to interpreter
-  bool    _release_icholder;
+// A CompiledICData* is a helper object for the inline cache implementation.
```
```suggestion
// A CompiledICData is a helper object for the inline cache implementation.
```

dean-long commented on src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:
```diff
@@ -757,17 +753,9 @@ AdapterHandlerEntry* SharedRuntime::generate_i2c2i_adapters(MacroAssembler *masm
+    __ ldr(rmethod, Address(data, CompiledICData::speculated_method_offset()));
 
-    __ bind(ok);
-    // Method might have been compiled since the call site was patched to
-    // interpreted; if that is the case treat it as a miss so we can get
-    // the call site corrected.
```
This comment is still relevant, isn't it?

fisk replied:
Unfortunately it is. I think it should be done in the subsequent patch_callers_callsite that has this exact same purpose for code dispatching through the VEP. But I think it's outside of scope for this PR to change that. I'm putting the comment back.

dean-long commented on src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:
```diff
@@ -1538,24 +1526,13 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
-
-  __ far_jump(RuntimeAddress(SharedRuntime::get_ic_miss_stub()));
-
-  // Verified entry point must be aligned
```
Keep this comment?

fisk replied:
Yeah, will do.

dean-long commented on src/hotspot/cpu/arm/arm.ad:
```diff
@@ -869,26 +869,15 @@ uint BoxLockNode::size(PhaseRegAlloc *ra_) const {
-  __ load_klass(Rtemp, receiver);
-  __ cmp(Rtemp, iCache);
-  __ jump(SharedRuntime::get_ic_miss_stub(), relocInfo::runtime_call_type, noreg, ne);
+  __ ic_check(CodeEntryAlignment);
```
Do we care about CodeEntryAlignment here if the old code didn't?

fisk replied:
Perhaps not, but it seems weird not to align. But maybe I should use InteriorEntryAlignment here, so it's symmetric to AArch64.

dean-long commented on src/hotspot/cpu/arm/sharedRuntime_arm.cpp:
```diff
@@ -819,21 +815,12 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
-  __ b(verified, eq); // jump over alignment no-ops too
-  __ jump(SharedRuntime::get_ic_miss_stub(), relocInfo::runtime_call_type, Rtemp);
-  __ align(CodeEntryAlignment);
-
```
Keep the CodeEntryAlignment and VEP comment?

dean-long commented on src/hotspot/cpu/x86/macroAssembler_x86.cpp:
```diff
@@ -1341,13 +1342,41 @@ void MacroAssembler::ic_call(address entry, jint method_index) {
+  Register receiver = LP64_ONLY(j_rarg0) NOT_LP64(rcx);
+  Register data = rax;
+  Register temp = LP64_ONLY(rscratch1) NOT_LP64(rbx);
+
```
It would be nice if VerifyOops check could be added back, maybe as a follow-up RFE?

fisk replied:
Good idea.

dean-long commented on src/hotspot/cpu/x86/sharedRuntime_x86_32.cpp:
```diff
@@ -1449,23 +1441,11 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
-
-  __ jump(RuntimeAddress(SharedRuntime::get_ic_miss_stub()));
-
-  // verified entry must be aligned for code patching.
```
This comment still seems relevant.

fisk replied:
Good point, resurrecting the comment.

dean-long commented on src/hotspot/cpu/x86/x86_32.ad:
```diff
@@ -1383,24 +1383,12 @@ void MachUEPNode::format( PhaseRegAlloc *ra_, outputStream* st ) const {
-  masm.nop(nops_cnt);
-
-  assert(cbuf.insts_size() - insts_size == size(ra_), "checking code size of inline cache node");
+  masm.ic_check(CodeEntryAlignment);
```
I think there may still be a problem with OptoBreakpoint messing up the alignment.  I'll try to reproduce it on x64.

dean-long commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -1472,40 +1472,19 @@ void MachUEPNode::format(PhaseRegAlloc* ra_, outputStream* st) const
-  nops_cnt &= 0x3; // Do not add nops if code is aligned.
-  if (nops_cnt > 0)
-    masm.nop(nops_cnt);
+  masm.ic_check(CodeEntryAlignment);
```
I'm concerned about OptoBreakpoint and friends messing up the alignment:
https://github.com/openjdk/jdk/blob/10beb3184e14e2714cd836029680a8b2d3fd5011/src/hotspot/share/opto/output.cpp#L317

dean-long replied:
It's fine, the breakpoint is insert *after* the prologue.

fisk replied:
Yeah, that did have me confused as well.

dean-long commented on src/hotspot/share/c1/c1_LIRAssembler.cpp:
```diff
@@ -606,13 +606,14 @@ void LIR_Assembler::emit_op0(LIR_Op0* op) {
+    case lir_std_entry: {
       // init offsets
       offsets()->set_value(CodeOffsets::OSR_Entry, _masm->offset());
-      _masm->align(CodeEntryAlignment);
```
So we used to unconditionally align here, but never set CodeOffsets::Entry, meaning we got the default offset of 0 and the entry had to execute the alignment NOPs.

fisk replied:
Yes. I found those nops to sometimes be expensive. So in the new model, it's a bug for the VEP to not already be aligned, from the UEP, so we can skip executing the padding nops.

dean-long commented on src/hotspot/share/runtime/vmStructs.cpp:
```diff
@@ -211,8 +211,10 @@
+  volatile_nonstatic_field(CompiledICData,     _speculated_method,                            Method*)                               \
+  volatile_nonstatic_field(CompiledICData,     _speculated_klass,                             uintptr_t)                             \
+  nonstatic_field(CompiledICData,              _itable_defc_klass,                            Klass*)                                \
+  nonstatic_field(CompiledICData,              _itable_refc_klass,                            Klass*)                                \
```
I don't think it makes sense to export these fields and types until SA is changed to make use of them.

fisk replied:
Good point. I thought at least Graal needs them, but they have their own VM structs file where they are exposed. I'll remove this.

fisk commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -1834,34 +1605,11 @@ methodHandle SharedRuntime::handle_ic_miss_helper(TRAPS) {
   CodeBlob* cb = caller_frame.cb();
   CompiledMethod* caller_nm = cb->as_compiled_method();
 
-  for (;;) {
```
Good catch. I'll remove the comment.

fisk commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -1912,14 +1658,8 @@ methodHandle SharedRuntime::reresolve_call_site(TRAPS) {
+    address call_addr = caller_nm->call_instruction_address(pc);
 
     // Check relocations for the matching call to 1) avoid false positives,
     // and 2) determine the type.
```
Yeah I don't really understand what the comment is trying to say at all. I'll remove it.

fisk commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -2079,65 +1777,42 @@ JRT_LEAF(void, SharedRuntime::fixup_callers_callsite(Method* method, address cal
+    return;
   }
+
+  assert(iter.has_current(), "must have a reloc at java call site");
```
Good point. Removing.

fisk commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -1464,39 +1347,33 @@ methodHandle SharedRuntime::resolve_sub_helper(bool is_virtual, bool is_optimize
+    inline_cache->update(&call_info, receiver->klass());
+  } else {
+    // Callsite is a direct call - set it to the destination method
+    CompiledICLocker ml(caller_nm);
```
Sure, I can do that.

fisk commented on src/hotspot/share/oops/oopsHierarchy.hpp:
```diff
@@ -180,8 +180,6 @@ class   MethodData;
 class   Method;
 class   ConstantPool;
 //      class CHeapObj
```
Okay, removing.

fisk commented on src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:
```diff
@@ -293,27 +292,7 @@ void LIR_Assembler::osr_entry() {
-
-  __ bind(dont);
-  return start_offset;
+  return __ ic_check(CodeEntryAlignment);
```
It's not obvious to me how likely this optimization is to kick in, and if it does, how many bytes it really saves. It optimizes accessor methods, when the UEP can be squeezed down to 4 instructions. The previous inline_cache_check was >= 2 instructions, there is a jump to skip the far jump, and the far_jump is >= 1 instructions. So it seems like it would kick in when far jumps can just branch and the compressed class encoding is simple enough. However, with the new ic_check we always have at least 5 instructions, sometimes 7. So it seemed to me like the intended optimization wouldn't apply any longer anyway.

fisk commented:
Thanks for the reviews @tschatzl and @dean-long. I have pushed changes reflecting your feedback.

dean-long commented on src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:
```diff
@@ -1571,6 +1551,8 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
     __ clinit_barrier(rscratch2, rscratch1, &L_skip_barrier);
     __ far_jump(RuntimeAddress(SharedRuntime::get_handle_wrong_method_stub()));
 
+    // Verified entry point must be aligned
```
I think this comment should come around line 1539.

fisk replied:
Okay, I will fix it.

dean-long commented on src/hotspot/cpu/arm/sharedRuntime_arm.cpp:
```diff
@@ -819,21 +815,13 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
-
-  // Verified entry point
-  __ bind(verified);
   int vep_offset = __ pc() - start;
```
Maybe add back
`// Verified entry point`
above this line.

fisk replied:
Sure.

dean-long commented:
If I just look at the new code, everything looks very reasonable.  I tried to compare the new code to the old code, but quickly gave up.

Could you explain why opt_virtual can now be a direct call and CompiledIC is now only for virtual?  It seems like we could have done that even with the old code.

Also, why don't we have to check for method->is_old() anymore?

dean-long commented on src/hotspot/share/code/compiledIC.cpp:
```diff
@@ -75,593 +67,327 @@ bool CompiledICLocker::is_safe(address code) {
-}
+void CompiledDirectCall::set(const methodHandle& callee_method) {
+  CompiledMethod* code = callee_method->code();
+  CompiledMethod* caller = CodeCache::find_compiled(instruction_address());
```
Instead of doing the slow `find_compiled` unconditionally, I think it would be better to check `is_interpreted_call` first.

dean-long commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -1219,29 +1238,29 @@ void nmethod::verify_clean_inline_caches() {
+        if (nm != nullptr) {
           // Verify that inline caches pointing to bad nmethods are clean
-          if (!nm->is_in_use() || (nm->method()->code() != nm)) {
+          if (!nm->is_in_use() || nm->is_unloading()) {
```
Please explain this change.  It's not obvious.

dean-long commented on src/hotspot/share/gc/x/xUnload.cpp:
```diff
@@ -101,7 +101,7 @@ class XCompiledICProtectionBehaviour : public CompiledICProtectionBehaviour {
 
   virtual bool is_safe(CompiledMethod* method) {
-    if (SafepointSynchronize::is_at_safepoint()) {
+    if (SafepointSynchronize::is_at_safepoint() || method->is_unloading()) {
```
Please explain why is_unloading() is needed now.

dean-long commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -2079,65 +1771,41 @@ JRT_LEAF(void, SharedRuntime::fixup_callers_callsite(Method* method, address cal
 
   CodeBlob* cb = CodeCache::find_blob(caller_pc);
-  if (cb == nullptr || !cb->is_compiled() || callee->is_unloading()) {
+  if (cb == nullptr || !cb->is_compiled() || !callee->is_in_use() || callee->is_unloading()) {
```
Checking for is_in_use() is just an optimization, right?

dean-long commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -2079,65 +1771,41 @@ JRT_LEAF(void, SharedRuntime::fixup_callers_callsite(Method* method, address cal
+  }
+
+  CompiledDirectCall* callsite = CompiledDirectCall::before(return_pc);
+  callsite->set_to_clean();
```
Clever way to get rid of should_fixup_call_destination(), something I've wanted to do!

fisk commented:
> If I just look at the new code, everything looks very reasonable. I tried to compare the new code to the old code, but quickly gave up.

:)

> Could you explain why opt_virtual can now be a direct call and CompiledIC is now only for virtual? It seems like we could have done that even with the old code.

Sure! So I think at some point we had static calls and virtual calls, with virtual calls using CompiledIC and static calls using CompiledStaticCall. I think there must have been some confusion at some point when introducing opt_virtual. They are virtual calls, so it might have seem "natural" to jam them into CompiledIC, and have CompiledIC deal with them. But since they in their code shapes are a lot more similar to static calls, with a direct call and a stub for interpreter entry, it never was a great fit. To deal with it we had to 1) ensure the data is always null as there is no data, because it isn't really an inline cache, 2) check for it everywhere to not spin up ICStubs when transitioning, because it isn't really an inline cache, and then when manipulating the callsite, have some some native call wrapper abstraction with virtual calls that would convert requests to update the inline cache to update the corresponding direct call and stub instead for optimized virtual calls, and get out of the inline cache world.

To me this always seemed a bit backwards. So in my new implementation I instead accept that opt_virtual calls really are more like the static calls (generated code is identical), and have pretty much nothing to do with inline caches, despite being virtual calls, and made them both use a common DirectCall abstraction instead, that fit both of them, as they are both direct calls.

Could we have cleaned that up earlier? Yes, probably. But it was pretty ingrained and "interesting" to change with incremental changes. I figured this was my chance to do this right since I'm rewriting the CompiledIC file pretty much from scratch as you noticed. I hope you agree with this decision.

> Also, why don't we have to check for method->is_old() anymore?

The checks for is_old were there because when performing an inline cache transition, we had situations where we would need an ICStub, and after running out of ICStubs we would have to request a safepoint to refill ICStubs. That safepoint could sneak in a class redefinition operation, rendering the methods invalid, or at least is_old(). Since I removed ICStubs and don't need any ICStubs to transition inline caches, we also get any safepoints in this code. And therefore we can't get any class redefinition, and hence can remove all the is_old() checks, which are now effectively dead code.

fisk commented:
Thanks for the review @dean-long! I updated the comments as requested.

fisk commented:
Thanks for the review, @dean-long! I pushed a trivial merge conflict with the next round of NULL -> nullptr changes. Running final round of tests now

fisk commented:
Tier1-7 test results look good.

fisk commented:
/integrate

fisk commented:
Thanks for the review @tschatzl!
